### PR TITLE
balance: refactor anticipation limits

### LINKED ISSUE #1084: Extrato - Limite não é atualizado corretamente
### Contexto
Ao acessar o formulário de antecipação, e então a partir dele acessar a página de extrato, o limite antecipável é apresentado de forma sempre zerada, exigindo que façamos o reload da página.

### Exemplo
peek 2019-02-05 14-39

### Requisitos e Critérios de Aceite
- [ ] Limite disponível para antecipação é recalculado ao acessar de qualquer lugar a página de extrato

### PR DESCRIPTION
**Pull request by alalbux**
## Contexto
Ao acessar o formulário de antecipação, e então a partir dele acessar a página de extrato, o limite antecipável é apresentado de forma sempre zerada, exigindo que façamos o reload da página.

## Checklist
- [ ] Dados da antecipação no epic do redux
- [ ] Adicionar action no Balance

## Issues linkadas
- [ ] Resolves #1084

### CODE REVIEW

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 663d091e658b565145e70453d002dea97fc24a80
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 663d091e658b565145e70453d002dea97fc24a80
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

File: packages/pilot/src/pages/Anticipation/epic.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    recipientId,
+  })
+))
+
+const limitsEpic = (action$, state$) =>
+  action$
+    .pipe(
+      ofType(ANTICIPABLE_LIMITS_REQUEST),
+      mergeMap(({ error, payload }) => {
+        if (error) {
```
Comment (rsmelo) [inline packages/pilot/src/pages/Anticipation/epic.js]: em que casos pode existir um `error` aqui? quando tiver erro nas validações do form?

File: packages/pilot/src/pages/Anticipation/reducer.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      }
+    }
+
+    case ANTICIPABLE_LIMITS_RECEIVE: {
+      const {
+        error,
+        payload,
+      } = action
+
+      if (error) {
```
Comment (rsmelo) [inline packages/pilot/src/pages/Anticipation/reducer.js]: poderia existir uma action `ANTICIPABLE_LIMITS_FAIL` para falhas de request a ser tratada com um ctach no epico como é feito no login, sem a necessidade desse if e deixando o nome das ações mais especificos https://github.com/pagarme/pilot/blob/d1824a2f995bccf05165ea137bf08bf4af259771/packages/pilot/src/pages/Account/actions/epic.js#L50

File: packages/pilot/src/pages/Anticipation/epic.js (hunk 1 above)
Comment (miguelssrs) [reply packages/pilot/src/pages/Anticipation/epic.js]: ☝️ @vagnervst @alalbux

File: packages/pilot/src/pages/Balance/Balance.js [hunk 3]
```diff
@@ -590,6 +572,11 @@ class Balance extends Component {
 }
 
 Balance.propTypes = {
+  anticipation: PropTypes.shape({
+    available: PropTypes.number.isRequired,
+    error: PropTypes.bool.isRequired,
```
Comment (miguelssrs) [inline packages/pilot/src/pages/Balance/Balance.js]: Aparentemente essa prop `error` está voltando como `undefined` e reclamando no console por estar required.
Screen Shot 2019-03-14 at 12 42 38

**Comment by rsmelo** [comment]
após fazer uma antecipação, ao retornar pra pagina de extrato as actions de  `ANTICIPABLE_LIMITS_REQUEST` e `ANTICIPABLE_LIMITS_RECEIVE` são disparadas inumeras vezes

Screen Shot 2019-03-14 at 20 26 58

File: packages/pilot/src/pages/Balance/Balance.js (hunk 3 above)
Comment (vagnervst) [reply packages/pilot/src/pages/Balance/Balance.js]: Boa! Corrigido!

File: packages/pilot/src/pages/Balance/Balance.js [hunk 4]
```diff
@@ -590,6 +562,11 @@ class Balance extends Component {
...
+    loading: PropTypes.bool.isRequired,
```
[the remark this answers is not in the document]
Comment (vagnervst) [reply packages/pilot/src/pages/Balance/Balance.js]: Corrigido, elas são utilizadas agora

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 3c4b92ad1ab9578c94bf111d9055561b6a1e20e2
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 3c4b92ad1ab9578c94bf111d9055561b6a1e20e2
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 3c4b92ad1ab9578c94bf111d9055561b6a1e20e2
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 3c4b92ad1ab9578c94bf111d9055561b6a1e20e2
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 3c4b92ad1ab9578c94bf111d9055561b6a1e20e2
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 3c4b92ad1ab9578c94bf111d9055561b6a1e20e2
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by rsmelo** [comment]
quando apaga o valor solicitado todo ele é subistiuido pelo valor maximo

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 2754de75e3ae3f729f102ab6fd79f7496c5ccf79
  * Good titles doesn't exceed 50 characters. To add more details use the body!

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There are some issues with your commit and our naming convention

 * Commit: 2754de75e3ae3f729f102ab6fd79f7496c5ccf79
  * Good titles doesn't exceed 50 characters. To add more details use the body!

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by rsmelo** [comment]
no carregamento inicial aparece `NaN` no total

File: packages/pilot/src/pages/Anticipation/Anticipation.js [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+      limits: {
+        max,
+        min,
+      },
+      loading: limitsLoading,
+      t,
+    } = this.props
+
+    const totalCost = -(anticipation + fraud + otherFee)
+    const amount = approximateRequested + totalCost + transferCost
```
Comment (rsmelo) [inline packages/pilot/src/pages/Anticipation/Anticipation.js]: acredito que isso aqui ajuda a resolver o problema de `NaN` na inicialização da pagina 
```suggestion
    const totalCostAndTransfer = totalCost + transferCost
    const amount = approximateRequested
      ? approximateRequested + totalCostAndTransfer
      : totalCostAndTransfer
```

Comment (vagnervst) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: Boa! Corrigi porém de outra forma, achei mais simples adicionar valores default na local state do component, assim `approximateRequested` e `transferCost` serão sempre números válidos, evitando o NaN

Comment (rsmelo) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: percebi que com essa validação acabou quebrando o fluxo de calcular os limites pois essa validação aqui não triga https://github.com/pagarme/pilot/blob/88920b3d899901e92e1b60d4e8a0b7cbffa7e16c/packages/pilot/src/pages/Anticipation/Anticipation.js#L337

Comment (vagnervst) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: Eita, verdade! Corrigi para manter da forma como tinha sugerido então

File: packages/pilot/src/containers/Anticipation/index.js [hunk 6]
```diff
@@ -137,26 +159,75 @@ class Anticipation extends Component {
       onAnticipationDateConfirm,
```
Comment (miguelssrs) [inline packages/pilot/src/containers/Anticipation/index.js]: essa prop tá sendo marcada como `required` aqui, porém está retornando `undefined` e dando um warningzin!

File: packages/pilot/src/pages/Balance/Balance.js [hunk 7]
```diff
@@ -590,6 +573,11 @@ class Balance extends Component {
 }
 
 Balance.propTypes = {
+  anticipation: PropTypes.shape({
+    available: PropTypes.number.isRequired,
```
Comment (miguelssrs) [inline packages/pilot/src/pages/Balance/Balance.js]: Será que não vale colocar um DefaultProps para `null` invés de colocar `isRequired`? É que existe um tratamento para se quando for `null` triggar o `requestAnticipationLimits` e, desse jeito que está, tem um warning no console reclamando sobre retornar `null`

File: packages/pilot/src/containers/Balance/index.js [hunk 8]
```diff
@@ -430,7 +431,12 @@ class Balance extends Component {
                 }
                 amount={formatAmount(outcoming)}
                 detail={this.renderAnticipation()}
-                disabled={disabled || anticipationLoading || anticipationError}
+                disabled={
+                  disabled
+                  || anticipationLoading
+                  || anticipationError
+                  || available < 100
```
Comment (rsmelo) [inline packages/pilot/src/containers/Balance/index.js]: talvez seja interessante definir uma  const pra esse caso para deixar mais claro o que essa validação significa, o que acha?
exemplo
```js
const MINIMUM_ANTICIPABLE_VALUE = 100
```

File: packages/pilot/src/pages/Anticipation/Anticipation.js [hunk 9]
```diff
@@ -0,0 +1,914 @@
...
+])
+
+const stepsId = {
+  confirmation: 'confirmation',
+  data: 'data',
+  result: 'result',
+}
+
+const initialState = {
+  approximateRequested: undefined,
```
Comment (rsmelo) [inline packages/pilot/src/pages/Anticipation/Anticipation.js]: que tal mudar esse `undefined` para null visto que já estamos setando um valor?

File: packages/pilot/src/pages/Anticipation/Anticipation.js [hunk 10]
```diff
@@ -0,0 +1,914 @@
...
+  feesValues: {
+    anticipation: 0,
+    fraud: 0,
+    otherFee: 0,
+  },
+  invalidDays: [],
+  isAutomaticTransfer: true,
+  loading: false,
+  paymentDate: moment(),
+  requestedAmount: undefined,
```
Comment (rsmelo) [inline packages/pilot/src/pages/Anticipation/Anticipation.js]: que tal mudar esse `undefined` para null visto que já estamos setando um valor?

File: packages/pilot/src/pages/Balance/Balance.js (hunk 7 above)
Comment (rsmelo) [reply packages/pilot/src/pages/Balance/Balance.js]: no caso não precisa o defualt value pois é uma subprop de `anticipation`. Basta remover o `isRequired`

File: packages/pilot/src/pages/Anticipation/Anticipation.js [hunk 11]
```diff
@@ -0,0 +1,914 @@
...
+            amount={amount}
+            approximateRequested={approximateRequested}
+            automaticTransfer={isAutomaticTransfer}
+            currentStep={currentStep}
+            date={paymentDate}
+            error={error}
+            loading={loading || limitsLoading}
+            maximum={max}
+            minimum={getMinLimit(min)}
+            onAnticipationDateConfirm={this.handleDateConfirm}
```
Comment (rsmelo) [inline packages/pilot/src/pages/Anticipation/Anticipation.js]: a função ```this.handleDateConfirm``` não é definida, me parece que essa prop `onAnticipationDateConfirm` se tornou desnecessária com a nova versão de `DateInput`. Visto que ela é passada pela arvore até o `Form` em todos os niveis. Inclusive isso deve resolver o que foi comentando nesse item tambem

Comment (rsmelo) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: apliquei essa alteração

File: packages/pilot/src/pages/Balance/Balance.js (hunk 7 above)
Comment (rsmelo) [reply packages/pilot/src/pages/Balance/Balance.js]: apliquei a correção para available não ser mais required

File: packages/pilot/src/pages/Anticipation/Anticipation.js (hunk 10 above)
Comment (rsmelo) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: resolvido

File: packages/pilot/src/pages/Anticipation/Anticipation.js (hunk 9 above)
Comment (rsmelo) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: resolvido

File: packages/pilot/src/containers/Balance/index.js (hunk 8 above)
Comment (rsmelo) [reply packages/pilot/src/containers/Balance/index.js]: resovido

File: packages/pilot/src/containers/Anticipation/index.js (hunk 6 above)
Comment (rsmelo) [reply packages/pilot/src/containers/Anticipation/index.js]: essa prop não era mais necessária e foi removida

File: packages/pilot/src/containers/Anticipation/index.js [hunk 12]
```diff
@@ -331,7 +352,7 @@ Anticipation.propTypes = {
 }
 
 Anticipation.defaultProps = {
-  approximateRequested: null,
+  approximateRequested: 0,
```
Comment (miguelssrs) [inline packages/pilot/src/containers/Anticipation/index.js]: Essa prop ainda está retornando `null` e dando warning no console:

```
Warning: Failed prop type: The prop `approximateRequested` is marked as required in `AnticipationFormContainer`, but its value is `null`.
```

Comment (rsmelo) [reply packages/pilot/src/containers/Anticipation/index.js]: eita eu tinha quebrado quando tirei os `undefined` no state do componte ja reverti

File: packages/pilot/src/pages/Anticipation/Anticipation.js (hunk 9 above)
Comment (rsmelo) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: reverti, tem que ficar com undefined mesmo

File: packages/pilot/src/pages/Anticipation/Anticipation.js (hunk 10 above)
Comment (rsmelo) [reply packages/pilot/src/pages/Anticipation/Anticipation.js]: reverti, tem que ficar com undefined mesmo

**Pull request merged**